File: C++/373_FindKPairswithSmallestSums.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
       // Min-heap to store pairs {sum, index1, index2}
        priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, greater<>> minHeap;
        vector<vector<int>> result;

        // Push the first pair of each element in nums1 with nums2[0]
        for (int i = 0; i < nums1.size() && i < k; ++i) {
            minHeap.emplace(nums1[i] + nums2[0], i, 0);
        }

        // Extract the smallest pairs up to k times
        while (k-- > 0 && !minHeap.empty()) {
            auto [sum, i, j] = minHeap.top();
            minHeap.pop();
            result.push_back({nums1[i], nums2[j]});

            // If there's a next pair in the row, push it into the heap
            if (j + 1 < nums2.size()) {
                minHeap.emplace(nums1[i] + nums2[j + 1], i, j + 1);
            }
        }

        return result;

    }
};

File: Python3/373_FindKPairswithSmallestSums.py
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        hp = []

        for i in range(min(k, len(nums1))):
            heappush(hp, (nums1[i]+nums2[0], i, 0))

        ans = []

        while hp and len(ans) < k:
            _, i, j = heappop(hp)
            ans.append([nums1[i], nums2[j]])
            if j + 1 < len(nums2):
                heappush(hp, (nums1[i]+nums2[j+1],i,j+1))

        return ans
        